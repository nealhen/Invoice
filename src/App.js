import React from "react";
import Table from "./table";
import styled from "styled-components";

const StyledForm = styled.form`
  width: 100%;
  margin: 10px 0 0;
  display: inline-grid;
  grid-template-columns: 25% 25% 25% 25%;
`;

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: "",
      cost: 0,
      quantity: 0,
      data: [],
      sum: 0
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event) {
    const target = event.target.name;
    this.setState({ [target]: event.target.value });
  }

  handleSubmit(event) {
    event.preventDefault();
    // simple input validation
    if (this.state.description === "" || this.state.quantity === 0) {
      alert("please eneter a description and qauntity");
      return 0;
    }
    const row = {
      description: this.state.description,
      cost: this.state.cost,
      quantity: this.state.quantity,
      price: this.state.cost * this.state.quantity
    };

    var joined = [...this.state.data, row];
    this.setState({ data: joined });
    const summed = this.state.sum + row.price;
    this.setState({ sum: summed });
    // empty inputs after submit
    this.setState({ description: "", cost: 0, quantity: 0 });
  }

  submitInvoice() {
    console.log(this.state.data);
  }

  render() {
    return (
      <div className="App">
        <StyledForm onSubmit={this.handleSubmit}>
          <div>
            <label>
              Description:
              <input
                type="text"
                name="description"
                value={this.state.description}
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div>
            <label>
              Cost:
              <input
                type="number"
                name="cost"
                value={this.state.cost}
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div>
            <label>
              Quantity:
              <input
                type="number"
                name="quantity"
                value={this.state.quantity}
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div>
            <input type="submit" value="Add Item" />
          </div>
        </StyledForm>
        <Table data={this.state.data} />
        <div className="padding right-align">Total: ${this.state.sum}</div>
        <button
          className="padding right-align"
          onClick={() => {
            this.submitInvoice();
          }}
        >
          Submit Invoice
        </button>
      </div>
    );
  }
}

export default App;
