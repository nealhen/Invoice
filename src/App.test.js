import React from 'react'
import { fireEvent, render, screen } from '@testing-library/react'
import App from './App'


const setup = () => {
  const app = render(<App />);
  const descriptionInput = app.getByLabelText('Description:');
  const costInput = app.getByLabelText('Cost:')
  const quantityInput = app.getByLabelText('Quantity:');
  const submit = app.getByTestId('submit');
  return {
    descriptionInput,
    costInput,
    quantityInput,
    submit,
    ...app,
  }
}

// it('have text', () => {
//   render(<App />);
//   expect(screen.getByTestId("submit").textContent).toBe('Submit Invoice');
// });

it('should update the total when a row is added', () =>{
  const { descriptionInput, costInput, quantityInput, submit } = setup()
  fireEvent.change(descriptionInput, { target: { value: 'New House' } })
  fireEvent.change(costInput, { target: { value: 23 } })
  fireEvent.change(quantityInput, { target: { value: 3 } })
  fireEvent.submit(submit);

  expect(screen.getAllByTestId('row')[0].textContent).toBe('New House$233$69');
  expect(screen.getByTestId('total').textContent).toBe('Total: $69')
});

