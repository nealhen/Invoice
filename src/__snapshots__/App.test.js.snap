// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <form
      class="sc-bdfBwQ dMLVRY"
    >
      <div>
        <label>
          Description:
          <input
            name="description"
            type="text"
            value=""
          />
        </label>
      </div>
      <div>
        <label>
          Cost:
          <input
            name="cost"
            type="number"
            value="0"
          />
        </label>
      </div>
      <div>
        <label>
          Quantity:
          <input
            name="quantity"
            type="number"
            value="0"
          />
        </label>
      </div>
      <div>
        <input
          type="submit"
          value="Add Item"
        />
      </div>
    </form>
    <table>
      <tbody>
        <tr>
          <th>
            Description
          </th>
          <th>
            Cost
          </th>
          <th>
            Quantity
          </th>
          <th>
            Price
          </th>
        </tr>
      </tbody>
    </table>
    <div
      class="padding right-align"
    >
      Total: $0
    </div>
    <button
      class="padding right-align"
    >
      Submit Invoice
    </button>
  </div>
</DocumentFragment>
`;
